[DINESH1201] iP

DINESH1201 opened the pull request:
# INTRODUCING BOB 😄 

> Time is the scarcest resource and unless it is managed nothing else can be managed - _Peter Drucker_

Your ***personal assistant*** for your day to day tasks.

Here is what Bob can do for you
- [x] Take note of your tasks
- [x] Filter your tasks
- [ ] Sort by deadline (Coming soon)

Who is it for
- Students
- Teachers
- Adults
- Kids
- **EVERYONEEE**

Download Bob by doing the following steps
1. Download it
2. Run it
3. Add your tasks and **HAVE FUN**

---

If you want to customize Bob feel free to do. This is the `main` method for Bob.

```java
public static void main(String[] args) {

    Bob myBot = new Bob("./userdata.txt");
    myBot.startChatBot();

}
```

Visit my github at [DINESH1201](https://github.com/DINESH1201)

Commit:
Add Gradle support

Eclipse-Dominator committed:
Bump gradle and lib version

aureliony committed:
build.gradle: Prevent generating a second JAR file

In build.gradle, the dependencies on distZip and/or distTar causes
the shadowJar task to generate a second JAR file for which the
mainClass.set("seedu.duke.Duke") does not take effect.
Hence, this additional JAR file cannot be run.
For this product, there is no need to generate a second JAR file
to begin with.

Let's remove this dependency from the build.gradle to prevent the
shadowJar task from generating the extra JAR file.

DINESH1201 committed:
Level 0. Rename, Greet, Exit

DINESH1201 committed:
Level 1. Echo

DINESH1201 committed:
Level 2. Add, List

DINESH1201 committed:
Level 3. Mark as Done

DINESH1201 committed:
Level 4. ToDos, Events, Deadlines

DINESH1201 committed:
Automated Text UI Testing

DINESH1201 committed:
Level 5. Handle Errors

DINESH1201 committed:
Level 6. Delete

DINESH1201 committed:
Updated Printing

DINESH1201 committed:
Add save and load from data feature

After users adds tasks to their task list, the data is stored into userdata.txt
* After every command that the user uses, the data is written in the userdata.txt to ensure that all the changes are properly saved.
* If userdata.txt does not exist, it will be created.

On initial launch of the chatbot, data from the userdata.txt is loaded if applicable
* If userdata.txt does not exist, no entries will be loaded
* If there are corrupted entries, it will not be loaded and the user will be informed
* User will get information on how many entries were successfully loaded

DINESH1201 committed:
Add datetime for deadline and event tasks

Deadline tasks now require the /by field to be a date and could optionally include a time.

Event tasks now require the /from field and /to field to be a date and could optionally include a time.

DINESH1201 committed:
Remove extra print statement

DINESH1201 committed:
Add classes for Parser, Storage and TaskList

Created a new Storage class which handles
* Reading of data from userdata.txt
* Writing to userdata.txt

Created a new Parser class which handles
* Creating new Tasks (ToDos, Deadlines, EventTask)

Created a new TaskList class which handles
* adding of Task
* removing of Task at index
* marking Task as completed at index
* unmarking Task as incomplete at index
* exporting all the tasks

DINESH1201 committed:
Add UI and Command classes

UI class
* Takes care of all the UI elements of the chatbot (printing)
* Reads input from the user

Parser class
* Now parses the user command and returns a command class which can be executed later

Command classes
* Every command has its own respective command class now. Calling execute() will run the respective code for each command.

DINESH1201 committed:
Organize classes into packages

Reorganize the classes into packages

DINESH1201 committed:
Create Main class and repackage Bob

Create Main class where the code will be run from

Rename the package for better clarity and add Bob to the package

DINESH1201 committed:
Edit runtest files

DINESH1201 committed:
Update .gitignore

DINESH1201 committed:
Merge branch 'add-gradle-support'

DINESH1201 committed:
Remove Main class

Make use of Bob class to run the chat bot instead

Rename package to match standard naming convention

DINESH1201 committed:
Add test cases for Add, Delete and Bye

Add JUnit test cases to test
* AddCommand
* DeleteCommand
* ByeCommand

DINESH1201 committed:
Add unit test for command classes

Unit Test added for the following Command Classes
* DeleteCommand
* ListCommand
* MarkCommand
* UnmarkCommand

DINESH1201 committed:
Change TaskList to throw InvalidTaskNumberException directly

The following methods in TaskList throw InvalidTaskNumberException
* removeTaskAtIndex(int)
* markTaskAtIndex(int)
* unmarkTaskAtIndex(int)

DINESH1201 committed:
Add unit test for Tasks

Unit test for the following classes
* Deadline
* EventTask
* ToDo
* TaskList

DINESH1201 committed:
Update unit test for commands

Small changes to the assertions made

DINESH1201 committed:
Add unit test for parser class

Add unit test cases for parser class for all the methods

DINESH1201 committed:
Add find feature

DINESH1201 committed:
Add JavaDoc for all classes and methods

DINESH1201 committed:
Merge branch 'branch-A-JavaDoc'

DINESH1201 committed:
Add JavaDoc

Add JavaDoc for the following methods
* contains in Task
* filter in TaskList

DINESH1201 committed:
Update to follow coding standard

DINESH1201 committed:
Merge branch 'branch-A-CodingStandard'

DINESH1201 committed:
Fix merge conflicts

DINESH1201 committed:
Add checkstyle and update style

Add configuration for checkstyle

Edit files to match checkstyle

Shruthiiii03 commented on src/main/java/bob/Bob.java:
```diff
@@ -0,0 +1,75 @@
+    /**
+     * Starts the chatbot
+     */
+    public void startChatBot() {
```
I like how you made sure everything is encapsulated in its own classes, so it makes the code more readable and clean. All of the commands are defined in their own classes which is well done on your part!

Shruthiiii03 commented on src/main/java/bob/Parser.java:
```diff
@@ -0,0 +1,158 @@
+        Scanner scanner = new Scanner(userCommand);
+        String input = scanner.next();
+        Command command = switch (input) {
+        case "bye" -> new ByeCommand();
+        case "list" -> new ListCommand();
+        case "mark" -> new MarkCommand(scanner.nextInt() - 1);
+        case "unmark" -> new UnmarkCommand(scanner.nextInt() - 1);
+        case "todo" -> new AddTaskCommand(Parser.newToDo(scanner.nextLine().trim()));
+        case "deadline" -> new AddTaskCommand(Parser.newDeadline(scanner.nextLine().trim()));
+        case "event" -> new AddTaskCommand(Parser.newEvent(scanner.nextLine().trim()));
+        case "delete" -> new DeleteCommand(scanner.nextInt() - 1);
+        case "find" -> new FindCommand(scanner.nextLine().trim());
```
I found this to be pretty interesting and well done, having separate classes for each command really makes the code so much more cleaner and I think really brings out encapsulation. I've noticed that you did this in many of the other classes as well. Your code follows the OOP principles really well!

Shruthiiii03 commented on src/main/java/bob/Storage.java:
```diff
@@ -0,0 +1,81 @@
+                String in = scanner.nextLine();
+                try {
+                    Task newTask;
+                    switch (in.split(" ")[0]) {
```
Well-followed code style! I can see the proper use of Egyptian brackets and proper indentations as well.

Shruthiiii03 commented on src/main/java/bob/tasks/TaskList.java:
```diff
@@ -0,0 +1,123 @@
+     *
+     * @return String format of all the tasks in the taskList to be exported
+     */
+    public String export() {
```
I like how you used collections and streams to make your functions more concise and efficient.

Shruthiiii03 commented on src/test/java/bob/tasks/EventTaskTest.java:
```diff
@@ -0,0 +1,123 @@
+package bob.tasks;
```
I like how you did extremely comprehensive testing for all your classes and their methods!

Shruthiiii03 reviewed (commented):
Overall, I found your code extremely readable and clean, following the principles of OOP very well. It was very intuitive for me as I was reading your code and I like that you split your code into multiple classes to ensure encapsulation.

DINESH1201 committed:
Add basic chat functionality for GUI

Add basic chat functionality with fxml

TODO
Update Bob.getResponse to return a string when user enters a valid input

DINESH1201 committed:
Update to show reply from Bob

Update getResponse function to return response for each command

DINESH1201 committed:
Update format for GUI

Add padding for label

DINESH1201 committed:
Add assertions

Add assertions to Command, Task and Storage class

Asserting the value of input parameters to ensure that the input
follows the required standards for Strings and valid for other types

Assertions are done to ensure that during runtime the assumptions
of the input are verified.

DINESH1201 committed:
Update code quality

Edit javadoc to give better description of the methods

Separate long code for testing into multiple test cases

DINESH1201 committed:
Merge pull request #3 from DINESH1201/branch-A-Assertions

Add assertions

DINESH1201 committed:
Merge branch 'branch-A-CodeQuality'

DINESH1201 committed:
Add fuzzy search for find feature

Find command: use case-insensitive fuzzy matching

Find command is case-insensitive exact match only

A fuzzy match is more user-friendly if the user has a typo in their
find function

DINESH1201 committed:
Update Task.java

Add credits to original user

DINESH1201 committed:
Add user guide

DINESH1201 committed:
Add constraints to UI

Add contraints to UI elements so that
it readjusts based on window size

DINESH1201 committed:
Add error handling

Add error handling for empty arguements for the following commands
- mark
- unmark
- delete
- find

DINESH1201 committed:
Add UI screenshot

Add UI screenshot to readme

DINESH1201 committed:
Merge branch 'branch-UserGuide'

DINESH1201 committed:
Rename UI screenshot file

DINESH1201 committed:
Update Ui.png
